driver: ILI9340 LCD display driver

vanwinkeljan opened the pull request:
Minimal driver for ILI9340 LCD display driver including support
for adafruit 2.2" LCD display (1480)

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

dbkinder commented on samples/display/display.rst:
```diff
@@ -0,0 +1,10 @@
+.. _display-samples:
+
+Display Samples
+##############
```
underline of #### need to be at least as long as the title

dbkinder commented on samples/display/ili9340/README.rst:
```diff
@@ -0,0 +1,18 @@
+- `ILI9340 Display Driver`_
+
+
+.. _ILI9340 Display Driver`: https://www.adafruit.com/product/1480
```
delete the `` ` `` here

dbkinder commented on samples/display/ili9340/src/main.c:
```diff
@@ -0,0 +1,86 @@
+#define SCREEN_HEIGHT 240
+
+void main(void)
+{
```
Some comments would be nice...

dbkinder commented on samples/display/ili9340/README.rst:
```diff
@@ -0,0 +1,18 @@
+Overview
+********
+
+Sample application that alternates LCD corners between RGB.
```
I don't know what this means. (The corner pixels of the LCD will cycle from Red to Green to Blue every half-second?)  Please explain what the sample is showing on the display. Also, we have a template for samples (doc/samples/sample.tmpl) that you should follow to explain more about how to setup and build your sample and  mention this sample is for a specific board.

dbkinder commented on samples/display/display.rst:
```diff
@@ -0,0 +1,10 @@
+.. _display-samples:
```
you'll also need to edit samples/samples.rst to add this new display.rst to the index.

dbkinder reviewed (changes requested):
Thanks for starting the docs for the sample, but could use more information (as found in the samples template).

tbursztyka commented on drivers/display/ili9340/Kconfig:
```diff
@@ -0,0 +1,88 @@
+
+config ILI9340_DEV_NAME
+	string "ILI9340 device name"
+	depends on ILI9340
```
you could have an if ILI9340 ... endif # ILI9340 all around these config options, it would avoid adding depends on ILI9340.

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+void ili9340_transceive(struct ili9340_data *data, u8_t cmd, void *tx_data,
+			size_t tx_len, void *rx_data, size_t rx_len)
+{
+	struct spi_config spi_config = {
```
Doing so won't perform nicely. Here you'll force the SPI controller to reconfigure itself on each ili9340_transceive() call, as this spi_config pointer might vary (since it's allocated on the stack).

Better include your struct spi_config into your struct ili9340_data (remove the struct device spi pointer there, as spi_config has one). And initialize the struct only once. of course.

I see that you have 2 different frequencies for tx and rx. This changes would indeed need a different config.
However, as far as I can tell, as you never read from the display... just scrap the rx part. (remove the parametrs from ili9340_transceive() as well then)

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+	    .frequency = 15151515, /* max (write) speed 15MHz (66ns) */
+	    .operation = SPI_OP_MODE_MASTER | SPI_WORD_SET(8),
+	    .vendor = 0,
+	    .slave = 0,
```
You might want to have Kconfig option for the slave, depending where this display will be wired and how, users may need to set a different value there

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+	if ((tx_data != NULL) || (rx_data != NULL)) {
+		tx_buf.buf = tx_data;
+		tx_buf.len = tx_len;
+		struct spi_buf rx_buf = {.buf = rx_data, .len = rx_len};
```
always declare variable at the beginning of a code block, so this line should be moved 2 lines above

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+			/* max read speed 6.6MHZ (150ns) */
+			spi_config.frequency = 6666666;
+		}
+		spi_transceive(&spi_config, &tx_buf, 1, &rx_buf, 1);
```
taste issue, but try to aerate your code everywhere, have an empty line  after a } unless it's another } etc...

tbursztyka commented on drivers/display/ili9340/Kconfig:
```diff
@@ -0,0 +1,88 @@
+# Kconfig - ILI9340 display driver configuration options
```
don't create dedicated directory unless your really have a lot of dedicated files.

Here have your Kconfig as: 
drivers/display/Kconfig.ili9340

tbursztyka commented on drivers/display/ili9340/Makefile:
```diff
@@ -0,0 +1,2 @@
+obj-$(CONFIG_ILI9340) += ili9340.o
```
add these in drivers/display/Makefile

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+/* Copyright (c) 2017 dXplore
```
and finally drivers/display/display_ili9340.c

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+	 * device name.
+	 * So just set driver_api to 1 else the function call will fail
+	 */
+	dev->driver_api = (void *)1;
```
ouch.

Maybe it would be worth considering generalizing a bit through a display.h device driver API?

tbursztyka commented on include/drivers/display/ili9340.h:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @return 0 in case of no error
+ */
+int ili9340_init(struct device *dev);
```
do not expose device driver initialization function. DEVICE_INIT is using it properly already.

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+
+#define ILI9340_CMD_DATA_PIN_COMMAND 0
+#define ILI9340_CMD_DATA_PIN_DATA 1
+
```
move all these macros (and why not the struct ili9340_data too actually) into display_ili9340.h header file

tbursztyka commented on drivers/display/ili9340/adafruit_1480.c:
```diff
@@ -0,0 +1,109 @@
+#define ILI9340_DATA_PIXEL_FORMAT_RGB_18_BIT 0x60
+#define ILI9340_DATA_PIXEL_FORMAT_RGB_16_BIT 0x50
+#define ILI9340_DATA_PIXEL_FORMAT_MCU_18_BIT 0x06
+#define ILI9340_DATA_PIXEL_FORMAT_MCU_16_BIT 0x05
```
So, are these macros 100% specific to adafruit device, or are they generic to ili9340?
If it's the later, please move all these into display_ili9340.h

tbursztyka commented on drivers/display/ili9340/ili9340_lcd.h:
```diff
@@ -0,0 +1,31 @@
+/* Copyright (c) 2017 dXplore
```
drivers/display/display_ili9340.h

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+			 const u8_t r, const u8_t g, const u8_t b)
+{
+	SYS_LOG_DBG("Writing pixel");
+	u8_t rgb_data[] = {r, g, b};
```
variable declaration at a the beginning of code black, always. Apply that everywhere

tbursztyka commented on drivers/display/ili9340/ili9340.c:
```diff
@@ -0,0 +1,181 @@
+void ili9340_write_pixel(const struct device *dev, const u16_t x, const u16_t y,
+			 const u8_t r, const u8_t g, const u8_t b)
+{
+	SYS_LOG_DBG("Writing pixel");
```
Maybe you want a better message. Here it will write something like "ili9340_write_pixel: Writing pixel", sounds redundant to me. Instead you could print out the pixel references. x/y/r/g/b

Apply that comment everywhere, other debug message would gain some improvements.

tbursztyka commented on drivers/display/ili9340/Kconfig:
```diff
@@ -0,0 +1,88 @@
+	Specify the device name of the I2C master device to which ILI9340 is
+	connected.
+
+config SYS_LOG_ILI9340_LEVEL
```
That one you could generalize it in the main display Kconfig and call it SYS_LOG_DISPLAY_LEVEL

tbursztyka reviewed (dismissed):
Could you split your commit in 2?
- driver first
- sample in second

Also, your public API seems like it could be generalized, it's concise and looks proper to any display driver to me: write pixel, write_bitmap. 
on and off could go on the PM side of the API.

Could you propose an include/display.h device driver API?

vanwinkeljan committed:
driver: ILI9340 LCD display driver

Minimal driver for ILI9340 LCD display driver including support
for adafruit 2.2" LCD display (1480)

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan committed:
driver: sample: ILI9340 sample application

Sample application showing how to use the ILI9340 LCD display driver.

Signed-off-by: Jan Van Winkel <jan.van_winkel@dxplore.eu>

vanwinkeljan commented:
@dbkinder Documentation updated

@tbursztyka Took in account majority of the comments.
Exception is the generalization to a generic API. 
It is definitely a good idea to have a common API, but it I have te feeling it is petter to put it in a separate pull request.

dbkinder reviewed (changes requested):
Doc looks better, thanks. One typo in kconfig.

dbkinder commented:
...and some checkpatch errors. Check the shippable log.

SebastianBoe commented on samples/display/ili9340/src/Makefile:
```diff
@@ -0,0 +1 @@
+obj-y += main.o
```
This should be removed I assume.

dbkinder commented:
retest

nashif commented:
Please rebase and squash last commit, we should not be introducing Makefiles and then removing them.

vanwinkeljan commented:
@nashif rebased and squashed
